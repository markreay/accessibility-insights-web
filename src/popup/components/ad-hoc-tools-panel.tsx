// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Icon } from 'office-ui-fabric-react/lib/Icon';
import { Link } from 'office-ui-fabric-react/lib/Link';
import * as React from 'react';

import { DiagnosticViewToggleFactory } from './diagnostic-view-toggle-factory';

export interface AdHocToolsPanelProps {
    backLinkHandler: () => void;
    diagnosticViewToggleFactory: DiagnosticViewToggleFactory;
}

export class AdHocToolsPanel extends React.Component<AdHocToolsPanelProps, {}> {
    private sliceSize: number = 3;

    public render(): JSX.Element {
        const toggles = this.props.diagnosticViewToggleFactory.createTogglesForAdhocToolsPanel();
        const groups = this.getToggleGroups(toggles);

        return (
            <div className="ms-Grid main-section">
                <main>
                    <div className="ms-Grid-row">{groups}</div>
                </main>
                <div role="navigation" className="ad-hoc-tools-panel-footer">
                    <Link
                        className="insights-link"
                        onClick={this.props.backLinkHandler}
                        id="back-to-launchpad-link"
                    >
                        <Icon className="back-to-launch-pad-icon" iconName="back" />
                        Back to launch pad
                    </Link>
                </div>
            </div>
        );
    }

    private getToggleGroups(toggles: JSX.Element[]): JSX.Element[] {
        const groups: JSX.Element[] = [];

        while (toggles.length > 0) {
            const groupIndex = groups.length;
            const currentSlice = toggles.splice(0, this.sliceSize);
            this.insertDivider(currentSlice, groupIndex);

            const currentGroup: JSX.Element = (
                <div
                    key={`visualization-toggle-group-${groupIndex}`}
                    className={`ms-Grid-col ms-sm6 ms-md6 ms-lg6 ad-hoc-tools-panel-group-${groupIndex}`}
                >
                    {currentSlice}
                </div>
            );

            groups.push(currentGroup);
        }

        return groups;
    }

    private insertDivider(toggles: JSX.Element[], groupIndex: number): void {
        for (let index = 1; index < toggles.length; index += 2) {
            toggles.splice(
                index,
                0,
                <div
                    key={`diagnostic-view-toggle-divider-${groupIndex}-${index}`}
                    className="ms-fontColor-neutralLight launch-panel-hr"
                />,
            );
        }
    }
}
