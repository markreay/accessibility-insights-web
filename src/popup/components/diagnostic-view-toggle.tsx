// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link } from 'office-ui-fabric-react/lib/Link';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import { IToggle } from 'office-ui-fabric-react/lib/Toggle';
import * as React from 'react';
import { ContentLink, ContentLinkDeps } from 'views/content/content-link';
import { VisualizationToggle } from '../../common/components/visualization-toggle';
import { VisualizationConfiguration } from '../../common/configs/visualization-configuration';
import { VisualizationConfigurationFactory } from '../../common/configs/visualization-configuration-factory';
import { KeyCodeConstants } from '../../common/constants/keycode-constants';
import { TelemetryEventSource } from '../../common/extension-telemetry-events';
import { DetailsViewPivotType } from '../../common/types/details-view-pivot-type';
import { VisualizationStoreData } from '../../common/types/store-data/visualization-store-data';
import { VisualizationType } from '../../common/types/visualization-type';
import { DictionaryStringTo } from '../../types/common-types';
import { PopupActionMessageCreator } from '../actions/popup-action-message-creator';
import { DiagnosticViewClickHandler } from '../handlers/diagnostic-view-toggle-click-handler';

export interface DiagnosticViewToggleProps {
    deps: DiagnosticViewToggleDeps;
    featureFlags: DictionaryStringTo<boolean>;
    visualizationType: VisualizationType;
    visualizationConfigurationFactory: VisualizationConfigurationFactory;
    visualizationStoreData: VisualizationStoreData;
    actionMessageCreator: PopupActionMessageCreator;
    clickHandler: DiagnosticViewClickHandler;
    shortcutCommands: chrome.commands.Command[];
    telemetrySource: TelemetryEventSource;
    dom: Document;
}

export type DiagnosticViewToggleDeps = ContentLinkDeps;

export interface DiagnosticViewToggleState {
    isFocused: boolean;
}

export class DiagnosticViewToggle extends React.Component<
    DiagnosticViewToggleProps,
    DiagnosticViewToggleState
> {
    private configuration: VisualizationConfiguration;
    private toggle: React.RefObject<IToggle> = React.createRef<IToggle>();
    private dom: Document;
    private userEventListenerAdded: boolean;

    // Must be consistent with internal react naming for same property to work
    // tslint:disable-next-line: variable-name
    private _isMounted: boolean;

    constructor(props: DiagnosticViewToggleProps) {
        super(props);
        this.configuration = this.props.visualizationConfigurationFactory.getConfiguration(
            this.props.visualizationType,
        );
        this.dom = this.props.dom;
        this._isMounted = false;
        this.userEventListenerAdded = false;
        this.state = {
            isFocused: false,
        };
    }

    public render(): JSX.Element {
        const displayableData = this.configuration.displayableData;
        const shortcut = this.getCommandShortcut();
        return (
            <div>
                <div className="ms-Grid-row view-toggle-row">
                    <div className="ms-Grid-col ms-sm7">
                        <div className="ms-fontColor-neutralPrimary ms-fontWeight-semibold ms-fontSize-mPlus">
                            {displayableData.title}
                        </div>
                    </div>
                    <div className="ms-Grid-col ms-sm5 view-toggle-col" style={{ float: 'right' }}>
                        {this.renderToggleOrSpinner()}
                    </div>
                </div>
                <div className="ms-Grid-row">
                    <div className="ms-Grid-col ms-sm8">
                        {this.renderLink(displayableData.linkToDetailsViewText)}
                    </div>
                    <div className="ms-Grid-col ms-sm4 shortcut-label" style={{ float: 'right' }}>
                        <div className="ms-font-xs">{shortcut}</div>
                    </div>
                </div>
            </div>
        );
    }

    private renderToggleOrSpinner(): JSX.Element {
        const scanning = this.props.visualizationStoreData.scanning;
        const id = this.configuration.getIdentifier();
        const scanData = this.configuration.getStoreData(this.props.visualizationStoreData.tests);

        if (scanning === id) {
            return <Spinner size={SpinnerSize.small} componentRef={this.addUserEventListener} />;
        } else {
            const disabled = scanning != null;
            return (
                <VisualizationToggle
                    checked={scanData.enabled}
                    disabled={disabled}
                    onClick={ev =>
                        this.props.clickHandler.toggleVisualization(
                            this.props.visualizationStoreData,
                            this.props.visualizationType,
                            ev,
                        )
                    }
                    visualizationName={this.configuration.displayableData.title}
                    componentRef={this.toggle}
                    onFocus={this.onFocusHandler}
                    onBlur={this.onBlurHandler}
                />
            );
        }
    }

    public componentDidMount(): void {
        this._isMounted = true;
        this.setFocus();
    }

    public componentDidUpdate(): void {
        this.setFocus();
    }

    private setFocus(): void {
        if (this._isMounted && this.state.isFocused && this.toggle.current) {
            this.toggle.current.focus();
        }
    }

    private onFocusHandler = (): void => {
        if (this._isMounted) {
            this.setState({
                isFocused: true,
            });
        }
    };

    private onBlurHandler = (): void => {
        if (this._isMounted) {
            this.setState({
                isFocused: false,
            });
        }
    };

    private addUserEventListener = (): void => {
        if (!this.userEventListenerAdded) {
            this.dom.addEventListener('keydown', (event: any) => {
                if (event.keyCode === KeyCodeConstants.TAB) {
                    this.onBlurHandler();
                }
            });

            const hamburgerButton = this.dom.querySelector('.feedback-collapse-menu-button');
            if (hamburgerButton != null) {
                hamburgerButton.addEventListener('click', event => {
                    this.onBlurHandler();
                });
            }

            this.userEventListenerAdded = true;
        }
    };

    private renderLink(linkText: string): JSX.Element {
        if (this.configuration.guidance) {
            return (
                <ContentLink
                    deps={this.props.deps}
                    reference={this.configuration.guidance}
                    linkText={linkText}
                />
            );
        }

        return (
            <Link
                className="insights-link"
                href="#"
                onClick={ev =>
                    this.props.actionMessageCreator.openDetailsView(
                        ev,
                        this.props.visualizationType,
                        this.props.telemetrySource,
                        DetailsViewPivotType.fastPass,
                    )
                }
            >
                {linkText}
            </Link>
        );
    }

    private getCommandShortcut(): string {
        const commandName: string = this.configuration.chromeCommand;

        for (const command of this.props.shortcutCommands) {
            if (command.name === commandName) {
                return command.shortcut;
            }
        }

        throw new Error(`Cannot find command for name: ${commandName}`);
    }
}
