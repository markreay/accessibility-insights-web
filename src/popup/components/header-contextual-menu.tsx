// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { getRTL } from '@uifabric/utilities';
import {
    ContextualMenu,
    ContextualMenuItemType,
    DirectionalHint,
    IContextualMenuItem,
} from 'office-ui-fabric-react/lib/ContextualMenu';
import * as React from 'react';
import { TelemetryEventSource } from '../../common/extension-telemetry-events';
import { NamedFC } from '../../common/react/named-fc';
import { DetailsViewPivotType } from '../../common/types/details-view-pivot-type';
import { VisualizationType } from '../../common/types/visualization-type';
import { DictionaryStringTo } from '../../types/common-types';
import { PopupActionMessageCreator } from '../actions/popup-action-message-creator';
import { LaunchPanelHeaderClickHandler } from '../handlers/launch-panel-header-click-handler';
import { LaunchPanelHeader } from './launch-panel-header';

export type HeaderContextualMenuDeps = {
    popupActionMessageCreator: PopupActionMessageCreator;
    launchPanelHeaderClickHandler: LaunchPanelHeaderClickHandler;
};

export type HeaderContextualMenuProps = {
    deps: HeaderContextualMenuDeps;
    header: LaunchPanelHeader;
    popupWindow: Window;
    featureFlags: DictionaryStringTo<boolean>;
} & Pick<IContextualMenuItem, 'target'>;

const telemetryEventSource = TelemetryEventSource.HamburgerMenu;

export const HeaderContextualMenu = NamedFC<HeaderContextualMenuProps>(
    'HeaderContextualMenu',
    props => {
        const { deps, header, popupWindow } = props;
        const { popupActionMessageCreator, launchPanelHeaderClickHandler } = deps;

        const getItems = (): IContextualMenuItem[] => [
            {
                key: 'fast-pass',
                iconProps: {
                    iconName: 'Rocket',
                },
                onClick: ev =>
                    popupActionMessageCreator.openDetailsView(
                        ev,
                        VisualizationType.Issues,
                        telemetryEventSource,
                        DetailsViewPivotType.fastPass,
                    ),
                name: 'FastPass',
            },
            {
                key: 'assessment',
                iconProps: {
                    iconName: 'testBeakerSolid',
                },
                onClick: ev =>
                    popupActionMessageCreator.openDetailsView(
                        ev,
                        null,
                        telemetryEventSource,
                        DetailsViewPivotType.assessment,
                    ),
                name: 'Assessment',
            },
            {
                key: 'ad-hoc-tools',
                iconProps: {
                    iconName: 'Medical',
                },
                name: 'Ad hoc tools',
                onClick: () => launchPanelHeaderClickHandler.openAdhocToolsPanel(header),
            },
            {
                key: 'divider_1',
                itemType: ContextualMenuItemType.Divider,
            },
            {
                key: 'modify-shortcuts',
                name: 'Keyboard shortcuts',
                onClick: event => popupActionMessageCreator.openShortcutConfigureTab(event as any),
            },
            {
                key: 'help',
                iconProps: {
                    iconName: 'Unknown',
                },
                data: 'https://go.microsoft.com/fwlink/?linkid=2077937',
                onClick: (event, item) =>
                    launchPanelHeaderClickHandler.onClickLink(popupWindow, event as any, item),
                name: 'Help',
            },
        ];

        return (
            <ContextualMenu
                className="popup-menu"
                shouldFocusOnMount={true}
                target={props.target}
                onDismiss={event =>
                    launchPanelHeaderClickHandler.onDismissFeedbackMenu(header, event)
                }
                directionalHint={
                    getRTL() ? DirectionalHint.bottomRightEdge : DirectionalHint.bottomLeftEdge
                }
                items={getItems()}
            />
        );
    },
);
