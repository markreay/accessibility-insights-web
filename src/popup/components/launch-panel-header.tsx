// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { IContextualMenuItem } from 'office-ui-fabric-react/lib/ContextualMenu';
import * as React from 'react';
import { GearOptionsButtonComponent } from '../../common/components/gear-options-button-component';
import { DropdownClickHandler } from '../../common/dropdown-click-handler';
import { FeatureFlagStoreData } from '../../common/types/store-data/feature-flag-store-data';
import { PopupActionMessageCreator } from '../actions/popup-action-message-creator';
import { LaunchPanelHeaderClickHandler } from '../handlers/launch-panel-header-click-handler';
import { SupportLinkHandler } from '../support-link-handler';
import { Header } from './header';
import { HeaderContextualMenu, HeaderContextualMenuDeps } from './header-contextual-menu';

export type LaunchPanelHeaderDeps = {
    popupActionMessageCreator: PopupActionMessageCreator;
    dropdownClickHandler: DropdownClickHandler;
    launchPanelHeaderClickHandler: LaunchPanelHeaderClickHandler;
} & HeaderContextualMenuDeps;

export interface LaunchPanelHeaderProps {
    deps: LaunchPanelHeaderDeps;
    title: string;
    subtitle?: React.ReactChild;
    openAdhocToolsPanel: () => void;
    supportLinkHandler: SupportLinkHandler;
    popupWindow: Window;
    featureFlags: FeatureFlagStoreData;
}

export type LaunchPanelHeaderState = {
    isContextMenuVisible: boolean;
} & Pick<IContextualMenuItem, 'target'>;

export class LaunchPanelHeader extends React.Component<
    LaunchPanelHeaderProps,
    LaunchPanelHeaderState
> {
    constructor(props: LaunchPanelHeaderProps) {
        super(props);

        this.state = {
            isContextMenuVisible: false,
        };
    }

    public render(): JSX.Element {
        return (
            <Header
                title={this.props.title}
                subtitle={this.props.subtitle}
                rowExtraClassName="header-title"
            >
                {this.renderGearOptionsButton()}
            </Header>
        );
    }

    private renderGearOptionsButton(): JSX.Element {
        const { dropdownClickHandler, launchPanelHeaderClickHandler } = this.props.deps;

        return (
            <div className="ms-Grid-col ms-u-sm2 feedback-collapseMenuButton-col">
                <GearOptionsButtonComponent
                    dropdownClickHandler={dropdownClickHandler}
                    featureFlags={this.props.featureFlags}
                />
                <IconButton
                    iconProps={{ iconName: 'GlobalNavButton' }}
                    id="feedback-collapse-menu-button"
                    onClick={event =>
                        launchPanelHeaderClickHandler.onOpenContextualMenu(this, event)
                    }
                    ariaLabel="Help and Feedback menu"
                />
                {this.renderContextualMenu(this.state.isContextMenuVisible)}
            </div>
        );
    }

    public renderContextualMenu(isContextMenuVisible: boolean): JSX.Element {
        if (!isContextMenuVisible) {
            return null;
        }

        const { deps, popupWindow, featureFlags } = this.props;

        const props = {
            deps,
            popupWindow,
            featureFlags,
            header: this,
            target: this.state.target,
        };

        return <HeaderContextualMenu {...props} />;
    }
}
