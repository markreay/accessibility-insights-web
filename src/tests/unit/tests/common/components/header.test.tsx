// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';

import { Header, HeaderDeps } from 'common/components/header';

describe('Header', () => {
    it('renders per snapshot', () => {
        const deps = {
            applicationTitle: 'APPLICATION_TITLE',
        } as HeaderDeps;
        const wrapper = shallow(<Header deps={deps} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
