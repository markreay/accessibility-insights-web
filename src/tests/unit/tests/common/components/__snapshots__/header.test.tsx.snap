// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders per snapshot 1`] = `
<header
  className="headerBar"
>
  <Component
    deps={
      Object {
        "applicationTitle": "APPLICATION_TITLE",
      }
    }
  />
  <span
    className="headerTitle"
  >
    APPLICATION_TITLE
  </span>
  <div />
  <div
    className="spacer"
  />
  <div
    className="farItems"
  />
</header>
`;
